### PR TITLE
Finished work to allow Crucible.NPC to support custom customers and custom traders

### PR DESCRIPTION
**Pull request by AndrewFahlgren**
- Fixed 1.0 issue with custom icons
- Expanded existing code for Crucible.NPC customers
- Added traders to Crucible.NPC
- Added some basic skeleton work for Crucible.NPC factions

### CODE REVIEW

File: Crucible.NPCs/CrucibleNpcAppearanceConfig.cs [hunk 1]
```diff
@@ -64,23 +64,9 @@ public class CrucibleNpcAppearanceConfig
         /// <param name="npc">The NPC to apply the configuration to.</param>
         public void ApplyAppearance(CrucibleNpcTemplate npc)
         {
-            npc.Appearance.Clear();
```
Comment (AndrewFahlgren) [inline Crucible.NPCs/CrucibleNpcAppearanceConfig.cs:67]: Removed this clear to allow for people to copy most of an NPC while only specifying appearance elements they want to change.

File: Crucible.NPCs/Quest/CrucibleNPCDialogueQuestConfig.cs [hunk 2]
```diff
@@ -14,38 +14,29 @@
...
 
-            // TODO: Allow adding to start / middle / end / before npc / after npc
-            //CrucibleDayManager.BackfillDaysTo(this.Day);
-            //var count = CrucibleDayManager.GetNpcTemplatesForDay(this.Day).Count();
-            //CrucibleDayManager.InsertNpcTemplateToDay(this.Day, count, npcTemplate);
-        }
+        /// <summary>
+        /// Gets or sets the starting dialogue node.
+        /// </summary>
+        public CrucibleDialogueData.CrucibleDialogueNode Dialogue { get; set; }
```
Comment (AndrewFahlgren) [inline Crucible.NPCs/Quest/CrucibleNPCDialogueQuestConfig.cs:40]: I am not sure if this is within design guidelines but I went ahead and used some GameAPI classes in the config for dialogue. The dialogue system I designed uses a node concept which didn't translate well to traditional config classes. To remedy this the best solution may be to move the `CrucibleDialogueData.CrucibleDialogueNode` and `CrucibleDialogueData.CrucibleAnswerNode` classes into Crucible.NPC and rename them to look like config classes.

File: Crucible.NPCs/Quest/CrucibleQuestRequirementConfig.cs [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+    /// Configuration specifying an addional quest requirement.
+    /// </summary>
+    [DuckTypeCandidate(typeof(CrucibleAdditionalEffectsQuestRequirementConfig))]
+    [DuckTypeCandidate(typeof(CrucibleMainIngredientQuestRequirementConfig))]
+    [DuckTypeCandidate(typeof(CrucibleMaxIngredientsQuestRequirementConfig))]
+    [DuckTypeCandidate(typeof(CrucibleNeedIngredientQuestRequirementConfig))]
+    [DuckTypeCandidate(typeof(CrucibleNoIngredientQuestRequirementConfig))]
+    [DuckTypeCandidate(typeof(CrucibleStrongPotionQuestRequirementConfig))]
+    [DuckTypeCandidate(typeof(CrucibleWeakPotionQuestRequirementConfig))]
+    public abstract class CrucibleQuestRequirementConfig : CruciblePackageConfigNode
```
Comment (AndrewFahlgren) [inline Crucible.NPCs/Quest/CrucibleQuestRequirementConfig.cs:33]: This got a bit absurd but it ended up working pretty well. If you check out https://github.com/AndrewFahlgren/potioncraft-crucible/wiki/Crucible-Package-Reference#quest-requirements you can see it got a little complicated to explain. I am not sure if there is a better way to do this which may be more strait forward to the modder to implement.

File: Crucible.GameAPI/CrucibleQuest.cs [hunk 4]
```diff
@@ -0,0 +1,208 @@
...
+            get => this.Quest.desiredEffects.Select(e => e.name).ToList();
+            set => this.Quest.desiredEffects = value.Select(e => CruciblePotionEffect.GetPotionEffectByID(e))
+                                                    .Select(e => e.PotionEffect)
+                                                    .ToArray();
+        }
+
+        /// <summary>
+        /// Gets or sets the minimum and maximum chapter needed to encounter this quest.
+        /// </summary>
+        public (int, int) MinMaxChapters
```
Comment (RoboPhred) [inline Crucible.GameAPI/CrucibleQuest.cs:83]: Lets avoid tupple return types and just make 2 properties for MinChapter and MaxChapter.  Tupple work is a bit advanced for the skill set that this API is aimed at, and I am worried people wont know how to set just one.

Comment (RoboPhred) [reply Crucible.GameAPI/CrucibleQuest.cs:83]: Alternatively, keep the tupple property, but add MinChapter and MaxChapter properties alongside it.

File: Crucible.NPCs/Quest/CrucibleQuestRequirementConfig.cs (hunk 3 above)
Comment (RoboPhred) [reply Crucible.NPCs/Quest/CrucibleQuestRequirementConfig.cs:33]: The yaml api this creates looks pretty spot on to me, I think it's fine.

File: Crucible.NPCs/Trader/CrucibleInventoryItemSoldByNpcStaticConfig.cs [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+        public int ClosenessRequirement { get; set; }
+
+        /// <summary>
+        /// Applies all required configuration for this <see cref="CrucibleInventoryItemSoldByNpcStaticConfig"/>.
+        /// </summary>
+        /// <param name="subject">The trader template to apply this <see cref="CrucibleInventoryItemSoldByNpcStaticConfig"/> to.</param>
+        public void OnApplyConfiguration(CrucibleTraderNpcTemplate subject)
+        {
+            var inventoryItem = CrucibleInventoryItem.GetByName(this.InventoryItemName);
+            inventoryItem.ClosenessRequirement = this.ClosenessRequirement;
```
Comment (RoboPhred) [inline Crucible.NPCs/Trader/CrucibleInventoryItemSoldByNpcStaticConfig.cs]: This seems to set a closeness requirement that is set to the item globally regardless of trader.  Is this the case?

If so, we should not allow a per-trader setting if different traders will simply override the global value for all traders.

File: Crucible.GameAPI/CrucibleTraderNpcTemplate.cs [hunk 6]
```diff
@@ -47,6 +57,153 @@ internal CrucibleTraderNpcTemplate(NpcTemplate npcTemplate)
...
+                }
+
+                this.NpcTemplate.unlockAtChapter = value;
+            }
+        }
+
+        /// <summary>
+        /// Gets or sets the minimum and maximum karma range that this trader will spawn in.
+        /// </summary>
+        public (int, int) KarmaForSpawn
```
Comment (RoboPhred) [inline Crucible.GameAPI/CrucibleTraderNpcTemplate.cs]: Another tupple that should be broken out int a max and min prop

Comment (AndrewFahlgren) [reply Crucible.GameAPI/CrucibleTraderNpcTemplate.cs]: Added two properties for min/max.

File: Crucible.GameAPI/CrucibleQuest.cs (hunk 4 above)
Comment (AndrewFahlgren) [reply Crucible.GameAPI/CrucibleQuest.cs:83]: Added two properties for min/max.

File: Crucible.NPCs/Trader/CrucibleInventoryItemSoldByNpcStaticConfig.cs (hunk 5 above)
Comment (AndrewFahlgren) [reply Crucible.NPCs/Trader/CrucibleInventoryItemSoldByNpcStaticConfig.cs]: The closeness requirement actually is on a trader by trader basis. I left in the old closeness configuration option for ingredients which now acts a default closeness requirement. A closeness requirement can also be specified for each sold by configuration which will override the default if set.

**Comment by AndrewFahlgren** [comment]
I also added an examples directory to Crucible.NPC with a mod I made while testing to serve as an example to modders. Is there a better place for this?

**Comment by AndrewFahlgren** [comment]
Found a couple errors while testing NPC appearances more thoroughly. Updated PR with fixes and some updates to the test mod to include appearance information.

**Pull request merged**